70_ima_policy.md

```diff
@@ -0,0 +1,348 @@
+* Conversion to new JSON format with `keylime_convert_allowlist --allowlist ima_allowlist.txt --exclude exclude.txt --out ima_policy.json`
+* The conversion tool produces `ima_policy.json`
+* Adds a new agent with `keylime_tenant -c add --ima_policy ima.json ...`
+* Agent is added to attestation.
+
+#### Story 1b - Using the legacy flat file format with signature
+* User generates a allowlist `ima_allowlist.txt` and a exclude list
+  `exclude.txt`, allowlist signature `ima_allowlist.txt.sig` and the public key `allowlist.key`
+* Conversion to new JSON format with `keylime_convert_allowlist --allowlist ima_allowlist.txt --exclude exclude.txt --allowlist-sig ima_allowlist.txt.sig --allowlist-sig allowlist.key --out ima_policy.json`
+* The conversion tool validates the signature and produces `ima_policy.json`
+* Adds a new agent with `keylime_tenant -c add --ima_policy ima.json ...`
+* Agent is added to attestation.
+
+
+Note that in this case the signature is checked by the conversion tool and that
+the exclude list does not have a signature that can be validated. This is fine
+because the signature validation is currently only done locally.
+
+
+#### Story 2a - Using a JSON policy
+ * User generates a IMA JSON policy called `ima.json`
+ * Adds a new agent with `keylime_tenant -c add --ima_policy ima.json ...`
+ * Agent is added to attestation.
+
+#### Story 2b - Using a JSON policy with signature
+ * User generates a IMA JSON policy called `ima.json` and has a signature
+   `ima.json.sig` and a public key `policy.key`.
+ * Adds a new agent with `keylime_tenant -c add --ima_policy ima.json --ima-policy-sig ima.json.sig --ima-policy-sig-key policy.key ...`
+ * Agent is added to attestation.
+
+
+### Notes/Constraints/Caveats
+
+* This only affects the server side API of the verifier. Therefore we will not
+  implement backwards compatibility for older tenant versions.
+* Signature validation in the tenant will be only possible with the JSON schema.
+
+### Risks and Mitigations
+
+There are no new risks because we are just unifying the existing configuration
+options.
+
+## Design Details
+
+### New IMA policy JSON schema
+The IMA policy is formally specified as:
+```json
+{
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
+    "title": "Keylime IMA policy",
+    "type": "object",
+    "properties": {
+        "meta": {
+            "type": "object",
+            "properties": {
+                "version": {
+                    "type": "integer",
+                    "description": "Version number of the IMA policy schema"
+                }
+            },
+            "required": ["version"],
+            "additionalProperties": false
+        },
+        "release": {
+            "type": "number",
+            "title": "Release version",
+            "description": "Version of the IMA policy (arbitrarily chosen by the user)"
+        },
+        "hashes": {
```

Comment:
nit: I would use `digests` here (digest is the result of a hash function)

Reply:
Makes sense. I'll update that.

Reply:
Yes, good point. I just realised hashes is being consistent with the current key:

```json
{
    "release": "33.20210314.2.0",
    "meta": {
        "generator": "coreos-assembler-6f54fb72d8a11393a74a6513d199b7895c5030b4",
        "timestamp": "2021-03-16T19:01:46.592224",
        "image": "builds/33.20210314.2.0/x86_64/fedora-coreos-33.20210314.2.0-qemu.x86_64.qcow2"
    },
    "hashes": {
        "/usr/bin/[": [
            "1ccfb4980e7d2d85b9576b918aebdb5cb940c1cbb34de17dfb254b028a8f46f4"
        ],
        "/usr/bin/afterburn": [
            "8a6f344166d301258e629e6b8db4100880cdb17817a83c8348d663efbcc4796d"
        ],
        "/usr/sbin/rdsosreport": [
            "5dead78d6155a577e53d91e81002e54a125461f27e9a4859c77aecc6b52e0b37"
        ]
    }
}
```

Reply:
@THS-on looking at the above, I figure we ignore the other arbitrary fields (release / meta ,.,)

Reply:
> We'll probably have to accept both while we transition and issue a deprecation warning when hashes is used.

@mpeters I don't think that we need to support both. This is a new format. If we decide to switch to `digests` then we can do that now.

Reply:
Right, sorry I was thinking about the old allowlist format.